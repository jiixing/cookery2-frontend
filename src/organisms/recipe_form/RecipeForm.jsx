// @flow

import React from 'react'
import { connect } from 'react-redux'
import { createStructuredSelector } from 'reselect'

import { Form, Control } from 'react-redux-form'
import { FormGroup, Label, Input as BaseInput, Button } from 'reactstrap'

import { validSelector } from './selectors'
import { actionPicker } from 'state'
import { modelName, actions } from 'state/form'

import Input from 'organisms/input'

import type { Difficulty, RecipeForm as RecipeFormType } from 'types'
type Props = {
  valid: boolean,
  create: RecipeFormType => void,
}

// exported for testing
export class RecipeForm extends React.Component<Props> {
  static difficulties(): Array<Difficulty> {
    return ['easy', 'medium', 'difficult']
  }

  constructor() {
    super()

    this.handleSubmit = this.handleSubmit.bind(this)
  }

  handleSubmit(recipe: RecipeFormType) {
    this.props.create(recipe)
  }

  render() {
    const { valid } = this.props

    return (
      <Form model={modelName()} onSubmit={this.handleSubmit}>
        <Input
          name="name"
          type="text"
          placeholder="Awesome Recipe"
          validators={{ length: val => val && val.length > 0 }}
          messages={{
            length: 'Name cannot be empty',
          }}
        />

        <FormGroup>
          <Label for="difficulty">Difficulty</Label>
          <Control
            model=".difficulty"
            type="select"
            name="difficulty"
            id="difficulty"
            component={BaseInput}
          >
            {RecipeForm.difficulties().map(difficulty => (
              <option key={difficulty}>{difficulty}</option>
            ))}
          </Control>
        </FormGroup>

        <Input
          name="servings"
          type="number"
          validators={{
            value: val => val >= 1 && val <= 10,
          }}
          messages={{
            value: 'It should be between 1 and 10',
          }}
        />

        <Input
          name="duration"
          type="number"
          validators={{
            value: val => val >= 1 && val < 600,
          }}
          messages={{
            value: 'It should be between 1 and 600',
          }}
        />

        <Button
          color="primary"
          type="submit"
          className="float-right"
          disabled={!valid}
        >
          Create Recipe
        </Button>
      </Form>
    )
  }
}

export default connect(
  (state, props) =>
    createStructuredSelector({
      valid: validSelector,
    })(state, props),
  actionPicker(['create'])(actions)
)(RecipeForm)
