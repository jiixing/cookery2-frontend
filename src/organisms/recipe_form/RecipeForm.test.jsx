import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

import { RecipeForm } from './RecipeForm'

describe('components', () => {
  describe('RecipeForm', () => {
    it('renders correctly', () => {
      const component = shallow(<RecipeForm stars={3} />)
      expect(toJson(component)).toMatchSnapshot()
    })
  })
})
