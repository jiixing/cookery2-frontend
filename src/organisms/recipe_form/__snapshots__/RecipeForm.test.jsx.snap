// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components RecipeForm renders correctly 1`] = `
<Connect(Form)
  model="form.recipe"
  onSubmit={[Function]}
>
  <Connect(Input)
    messages={
      Object {
        "length": "Name cannot be empty",
      }
    }
    name="name"
    placeholder="Awesome Recipe"
    type="text"
    validators={
      Object {
        "length": [Function],
      }
    }
  />
  <FormGroup
    tag="div"
  >
    <Label
      for="difficulty"
      tag="label"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      Difficulty
    </Label>
    <DefaultConnectedControl
      component={[Function]}
      id="difficulty"
      model=".difficulty"
      name="difficulty"
      type="select"
    >
      <option
        key="easy"
      >
        easy
      </option>
      <option
        key="medium"
      >
        medium
      </option>
      <option
        key="difficult"
      >
        difficult
      </option>
    </DefaultConnectedControl>
  </FormGroup>
  <Connect(Input)
    messages={
      Object {
        "value": "It should be between 1 and 10",
      }
    }
    name="servings"
    type="number"
    validators={
      Object {
        "value": [Function],
      }
    }
  />
  <Connect(Input)
    messages={
      Object {
        "value": "It should be between 1 and 600",
      }
    }
    name="duration"
    type="number"
    validators={
      Object {
        "value": [Function],
      }
    }
  />
  <Button
    className="float-right"
    color="primary"
    tag="button"
    type="submit"
  >
    Create Recipe
  </Button>
</Connect(Form)>
`;
